huaweicloud/services/ddm/resource_huaweicloud_ddm_instance.go

```diff
@@ -0,0 +1,931 @@
+	)
+	deleteInstanceClient, err := cfg.NewServiceClient(deleteInstanceProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating DDM client: %s", err)
+	}
+
+	deleteInstancePath := deleteInstanceClient.Endpoint + deleteInstanceHttpUrl
+	deleteInstancePath = strings.ReplaceAll(deleteInstancePath, "{project_id}", deleteInstanceClient.ProjectID)
+	deleteInstancePath = strings.ReplaceAll(deleteInstancePath, "{instance_id}", fmt.Sprintf("%v", d.Id()))
+
+	deleteInstanceQueryParams := buildDeleteInstanceQueryParams(d)
+	deleteInstancePath += deleteInstanceQueryParams
+
+	deleteInstanceOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			200,
+		},
+		MoreHeaders: map[string]string{
+			"Content-Type": "application/json",
+		},
+	}
+	_, err = deleteInstanceClient.Request("DELETE", deleteInstancePath, &deleteInstanceOpt)
+	if err != nil {
+		return diag.Errorf("error deleting DdmInstance: %s", err)
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"RUNNING", "DELETING"},
+		Target:       []string{"DELETED"},
+		Refresh:      ddmInstanceStatusRefreshFunc(d.Id(), region, cfg),
+		Timeout:      d.Timeout(schema.TimeoutDelete),
+		Delay:        30 * time.Second,
+		PollInterval: 10 * time.Second,
+	}
+
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.Errorf("error waiting for instance (%s) to deleted: %s", d.Id(), err)
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func buildDeleteInstanceQueryParams(d *schema.ResourceData) string {
+	res := ""
+	if v, ok := d.GetOk("delete_rds_data"); ok {
+		res = fmt.Sprintf("%s&delete_rds_data=%v", res, v)
+	}
+
+	if res != "" {
+		res = "?" + res[1:]
+	}
+	return res
+}
+
+func ddmInstanceStatusRefreshFunc(id, region string, cfg *config.Config) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		var (
+			getJobStatusHttpUrl = "v1/{project_id}/instances/{instance_id}"
+			getInstanceProduct  = "ddm"
+		)
+
+		getInstanceClient, err := cfg.NewServiceClient(getInstanceProduct, region)
+		if err != nil {
+			return nil, "", fmt.Errorf("error creating DDM client: %s", err)
+		}
+
+		getJobStatusPath := getInstanceClient.Endpoint + getJobStatusHttpUrl
+		getJobStatusPath = strings.ReplaceAll(getJobStatusPath, "{project_id}", getInstanceClient.ProjectID)
+		getJobStatusPath = strings.ReplaceAll(getJobStatusPath, "{instance_id}", fmt.Sprintf("%v", id))
+
+		getJobStatusOpt := golangsdk.RequestOpts{
+			KeepResponseBody: true,
+			OkCodes: []int{
+				200,
+			},
+		}
+		getJobStatusResp, err := getInstanceClient.Request("GET", getJobStatusPath, &getJobStatusOpt)
+		if err != nil {
+			if _, ok := err.(golangsdk.ErrDefault400); ok {
+				return getJobStatusResp, "DELETED", nil
+			}
+			return nil, "", err
+		}
+
+		getJobStatusRespBody, err := utils.FlattenResponse(getJobStatusResp)
+		if err != nil {
+			return nil, "", err
+		}
+
+		status := utils.PathSearch("status", getJobStatusRespBody, nil)
```

Comment:
If the last param is `nil`, `status.(string)` may make panic error

Reply:
done